pkg/controller/pv_test.go

```diff
@@ -81,6 +83,119 @@ func TestPVControllerReconcile(t *testing.T) {
+				pv := testPersistentVolume0.DeepCopy()
+				pv.Status.Phase = corev1.VolumeReleased
+				pv.DeletionTimestamp = &metav1.Time{Time: time.Now().Add(time.Minute * -1)}
+
+				controller := newTestPVController(
+					mockCtl,
+					testNamespace,
+					pv)
+
+				mockClients(controller.cachedClient.(*mockclient.MockClient), controller.azClient, controller.kubeClient)
+
+				return controller
+			},
+			verifyFunc: func(t *testing.T, controller *ReconcilePV, result reconcile.Result, err error) {
+				require.NoError(t, err)
+				require.False(t, result.Requeue)
+			},
+		},
+		{
+			description: "[Success] Should delete AzVolume when PV is deleted but AzVolume exists.",
+			request:     testPersistentVolume0Request,
+			setupFunc: func(t *testing.T, mockCtl *gomock.Controller) *ReconcilePV {
+				azVolume := testAzVolume0.DeepCopy()
+
+				pv := testPersistentVolume0.DeepCopy()
+				pv.Status.Phase = corev1.VolumeReleased
+				pv.DeletionTimestamp = &metav1.Time{Time: time.Now().Add(time.Minute * -1)}
+
+				controller := newTestPVController(
+					mockCtl,
+					testNamespace,
+					azVolume,
+					pv)
+
+				mockClients(controller.cachedClient.(*mockclient.MockClient), controller.azClient, controller.kubeClient)
+
+				return controller
+			},
+			verifyFunc: func(t *testing.T, controller *ReconcilePV, result reconcile.Result, err error) {
+				require.NoError(t, err)
+				require.False(t, result.Requeue)
+				_, err = controller.azClient.DiskV1beta2().AzVolumes(controller.objectNamespace).Get(context.Background(), testPersistentVolume0Name, metav1.GetOptions{})
+				require.True(t, errors.IsNotFound(err))
+			},
+		},
+		{
+			description: "[Success] Should create AzVolume when PV exists but AzVolume isn't found.",
+			request:     testPersistentVolume0Request,
+			setupFunc: func(t *testing.T, mockCtl *gomock.Controller) *ReconcilePV {
+				pv := testPersistentVolume0.DeepCopy()
+				pv.Status.Phase = corev1.VolumeBound
+
+				controller := newTestPVController(
+					mockCtl,
+					testNamespace,
+					pv)
+
+				mockClients(controller.cachedClient.(*mockclient.MockClient), controller.azClient, controller.kubeClient)
+
+				return controller
+			},
+			verifyFunc: func(t *testing.T, controller *ReconcilePV, result reconcile.Result, err error) {
+				require.NoError(t, err)
+				require.False(t, result.Requeue)
+				azv, err := controller.azClient.DiskV1beta2().AzVolumes(controller.objectNamespace).Get(context.Background(), testPersistentVolume0Name, metav1.GetOptions{})
+				require.NoError(t, err)
+				require.Equal(t, "true", azv.Status.Annotations[consts.PreProvisionedVolumeAnnotation])
+			},
+		},
+		{
+			description: "[Success] Should ignore non-csi volumes",
+			request:     testPersistentVolume0Request,
+			setupFunc: func(t *testing.T, mockCtl *gomock.Controller) *ReconcilePV {
+				pv := testPersistentVolume0.DeepCopy()
+				pv.Status.Phase = corev1.VolumeBound
+				pv.Spec.CSI.Driver = "non-csi-driver"
```

Comment:
set `pv.Spec.CSI = nil` given that the test case calls for "non-csi" volumes?

Reply:
Yeah, make sense.

Reply:
Also edit `initState` function in `common_test.go` accordingly to prevent nil pointer dereference.